tests/auth_tests/test_models.py

```diff
@@ -261,6 +261,172 @@ def test_check_password_upgrade(self):
             hasher.iterations = old_iterations
 
 
+class UserWithPermTestCase(TestCase):
+
+    @classmethod
+    def setUpTestData(cls):
+        cls.content_type = ContentType.objects.get_for_model(Group)
+        cls.permission = Permission.objects.create(
+            name='test', content_type=cls.content_type, codename='test'
+        )
+        cls.user1 = User.objects.create_user('user1', 'foo@example.com')
+        cls.user1.user_permissions.add(cls.permission)
+
+        cls.group = Group.objects.create(name='test')
+        cls.group.permissions.add(cls.permission)
+        cls.group2 = Group.objects.create(name='test2')
+        cls.group2.permissions.add(cls.permission)
+        cls.user2 = User.objects.create_user('user2', 'bar@example.com')
+        cls.user2.groups.add(cls.group, cls.group2)
+
+        cls.user3 = User.objects.create_user('user3', 'baz@example.com')
+        cls.superuser = User.objects.create_superuser(
+            'superuser', 'superuser@example.com', 'superpassword',
+        )
+        cls.inactive_user = User.objects.create_user(
+            'inactive_user', 'baz@example.com', is_active=False,
+        )
+        cls.inactive_user.user_permissions.add(cls.permission)
+        cls.charlie = User.objects.create_user('charlie', 'charlie@example.com')
+        cls.charlie_brown = User.objects.create_user('charliebrown', 'charlie@brown.com')
+
+    def test_invalid_permission_name(self):
+        msg = "Permission name should be in the form 'app_label.perm_name'."
+        for perm in ('nodots', 'too.many.dots', '...', ''):
+            with self.assertRaisesMessage(ValueError, msg):
+                User.objects.with_perm(perm)
+
+    def test_user_permissions(self):
```

Comment:
This test is duplicated exactly in `test_group_permissions()` and `test_non_duplicate_users()`.